yt_dlp/extractor/nebula.py

```diff
@@ -73,10 +73,23 @@ def _retrieve_nebula_api_token(self):
     def _call_nebula_api(self, url, video_id=None, method='GET', auth_type='api', note=''):
         assert method in ('GET', 'POST',)
         assert auth_type in ('api', 'bearer',)
-        authorization = f'Token {self._nebula_api_token}' if auth_type == 'api' else f'Bearer {self._nebula_bearer_token}'
-        return self._download_json(
-            url, video_id, note=note, headers={'Authorization': authorization},
-            data=b'' if method == 'POST' else None)
+
+        def inner_call():
+            authorization = f'Token {self._nebula_api_token}' if auth_type == 'api' else f'Bearer {self._nebula_bearer_token}'
+            return self._download_json(
+                url, video_id, note=note, headers={'Authorization': authorization},
+                data=b'' if method == 'POST' else None)
+
+        try:
+            return inner_call()
+        except ExtractorError as exc:
+            # if 401 or 403, attempt credential re-auth and retry
+            if exc.cause and isinstance(exc.cause, urllib.error.HTTPError) and exc.cause.code in (401, 403):
+                self.to_screen(f'Reauthenticating to Nebula and retrying, because last {auth_type} call resulted in error {exc.cause.code}')
+                self._real_initialize()
+                return inner_call()
+            else:
+                raise
```

Comment:
another way to do this could be:
```py
for retry in range(2):
    try:
        return self._download_json(...)
    except ExtractorError as e:
        if not retry and ... and e.cause.code in (401, 403):
            self._login()
        else:
            raise
```

which is more consise, but maybe harder to reason about 🤔 ?

Reply:
Could we drop something like this into `utils.py`?

```
def _retry_once(main_func, recover_func, exception_classes):
    assert isinstance(exception_classes, tuple)

    for retry in range(2):
        try:
            return main_func()
        except exception_classes as e:
            if recover_func(e):
                return main_func()
            else:
                raise
```

Or is this overengineered...

Usage would look something like this (simplified, untested):

```
def inner_call():
    return self._download_json(url, video_id, ...)

def attempt_recover(exc):
    if exc.cause and isinstance(exc.cause, urllib.error.HTTPError) and exc.cause.code in (401, 403):
        self.to_screen(f'Reauthenticating to Nebula and retrying, because last {auth_type} call resulted in error {exc.cause.code}')
        self._login()
        return True

retry_once(inner_call, attempt_recover, exception_classes=(ExtractorError,))
```

I'll let this sit for a bit and think about it some more.

Reply:
Not sure how often this pattern is actually used. I'll see if this can be incorporated into https://github.com/yt-dlp/yt-dlp/pull/1649. Leave it be for now